extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerImpl.java

```diff
@@ -273,203 +189,112 @@ public TaskStatus run() throws Exception
-        criticalWarningCodes,
-        controllerClient,
-        id(),
-        MSQTasks.getHostFromSelfNode(selfDruidNode)
-    );
-
-    closer.register(msqWarningReportPublisher);
+    // Delay removal of kernels so we don't interfere with iteration of kernelHolders.getAllKernelHolders().
+    final Set<StageId> kernelsToRemove = new HashSet<>();
 
-    final Map<StageId, SettableFuture<ClusterByPartitions>> partitionBoundariesFutureMap = new HashMap<>();
-
-    final Map<StageId, FrameContext> stageFrameContexts = new HashMap<>();
-
-    while (!kernelHolder.isDone()) {
+    while (!kernelHolders.isDone()) {
       boolean didSomething = false;
 
-      for (final WorkerStageKernel kernel : kernelHolder.getStageKernelMap().values()) {
+      for (final KernelHolder kernelHolder : kernelHolders.getAllKernelHolders()) {
+        final WorkerStageKernel kernel = kernelHolder.kernel;
         final StageDefinition stageDefinition = kernel.getStageDefinition();
 
-        if (kernel.getPhase() == WorkerStagePhase.NEW) {
-
-          log.info("Processing work order for stage [%d]" +
-                   (log.isDebugEnabled()
-                    ? StringUtils.format(
-                       " with payload [%s]",
-                       context.jsonMapper().writeValueAsString(kernel.getWorkOrder())
-                   ) : ""), stageDefinition.getId().getStageNumber());
-
-          // Create separate inputChannelFactory per stage, because the list of tasks can grow between stages, and
-          // so we need to avoid the memoization in baseInputChannelFactory.
-          final InputChannelFactory inputChannelFactory = makeBaseInputChannelFactory(closer);
-
-          // Compute memory parameters for all stages, even ones that haven't been assigned yet, so we can fail-fast
-          // if some won't work. (We expect that all stages will get assigned to the same pool of workers.)
-          for (final StageDefinition stageDef : kernel.getWorkOrder().getQueryDefinition().getStageDefinitions()) {
-            stageFrameContexts.computeIfAbsent(
-                stageDef.getId(),
-                stageId -> context.frameContext(
-                    kernel.getWorkOrder().getQueryDefinition(),
-                    stageId.getStageNumber()
-                )
-            );
-          }
-
-          // Start working on this stage immediately.
-          kernel.startReading();
-
-          final RunWorkOrder runWorkOrder = new RunWorkOrder(
-              kernel,
-              inputChannelFactory,
-              stageCounters.computeIfAbsent(stageDefinition.getId(), ignored -> new CounterTracker()),
+        // Workers run all work orders they get. There is not (currently) any limit on the number of concurrent work
+        // orders; we rely on the controller to avoid overloading workers.
+        if (kernel.getPhase() == WorkerStagePhase.NEW
+            && kernelHolders.runningKernelCount() < context.maxConcurrentStages()) {
```

Comment:
Workers in `RESULTS_COMPLETE` running in the `MEMORY` mode would be consuming a decent portion of the memory storing the result output. Does this condition not overload the memory on the worker because at most 1 stage will be not-running but consuming memory (in leap-frogging)

Reply:
`MEMORY` just means the transfer is in-memory. The results aren't completely buffered in memory: just one frame at a time per channel. I added a note about this to `OutputChannelMode#MEMORY`.

Reply:
New text:

```
  /**
   * In-memory output channels. Stage shuffle data does not hit disk. In-memory channels do not fully buffer stage
   * output. They use a blocking queue; see {@link RunWorkOrder#makeStageOutputChannelFactory()}.
   *
   * Because stage output is not fully buffered, this mode requires a consumer stage to run at the same time as its
   * corresponding producer stage. See {@link ControllerQueryKernelUtils#computeStageGroups} for the logic that
   * determines when we can use in-memory channels.
   */
  MEMORY("memory"),
```